File: src/prompt/git.zig
//! Git status for the prompt. `parseStatus` (pure, tested) interprets the
//! output of `git status --porcelain=v1 --branch`; `query` runs git as a
//! subprocess and feeds it to the parser.

const std = @import("std");

pub const Info = struct {
    branch: []const u8, // borrowed from the caller-provided buffer
    dirty: u32 = 0,
    ahead: u32 = 0,
    behind: u32 = 0,
};

/// Parse `git status --porcelain=v1 --branch` output. The branch name is a
/// slice into `text`. Returns null if no branch header line is present.
pub fn parseStatus(text: []const u8) ?Info {
    var info: ?Info = null;
    var lines = std.mem.splitScalar(u8, text, '\n');
    while (lines.next()) |line| {
        if (line.len == 0) continue;
        if (std.mem.startsWith(u8, line, "## ")) {
            info = parseBranchLine(line[3..]);
        } else {
            if (info) |*i| i.dirty += 1;
        }
    }
    return info;
}

fn parseBranchLine(rest: []const u8) Info {
    // e.g. "main...origin/main [ahead 1, behind 2]"  or  "main"
    var branch_end: usize = rest.len;
    if (std.mem.indexOf(u8, rest, "...")) |i| branch_end = i;
    if (std.mem.indexOfScalar(u8, rest, ' ')) |i| branch_end = @min(branch_end, i);
    var info = Info{ .branch = rest[0..branch_end] };
    if (std.mem.indexOf(u8, rest, "ahead ")) |i| {
        info.ahead = readNum(rest[i + 6 ..]);
    }
    if (std.mem.indexOf(u8, rest, "behind ")) |i| {
        info.behind = readNum(rest[i + 7 ..]);
    }
    return info;
}

fn readNum(s: []const u8) u32 {
    var n: u32 = 0;
    for (s) |ch| {
        if (ch < '0' or ch > '9') break;
        n = n * 10 + (ch - '0');
    }
    return n;
}

/// Run git in `cwd` and return its status, or null if not a repo / git fails /
/// it errors. `out_buf` backs the returned branch slice.
///
/// Uses `std.process.run` (Zig 0.16 API) with a `Threaded` io backed by
/// `c_allocator` so that process spawn can allocate argv buffers.
pub fn query(allocator: std.mem.Allocator, cwd: []const u8, out_buf: []u8) ?Info {
    var threaded = std.Io.Threaded.init(std.heap.c_allocator, .{});
    const io = threaded.io();
    const result = std.process.run(allocator, io, .{
        .argv = &.{ "git", "status", "--porcelain=v1", "--branch" },
        .cwd = .{ .path = cwd },
    }) catch return null;
    defer allocator.free(result.stdout);
    defer allocator.free(result.stderr);

    switch (result.term) {
        .exited => |code| if (code != 0) return null,
        else => return null,
    }

    const parsed = parseStatus(result.stdout) orelse return null;
    if (parsed.branch.len > out_buf.len) return null;
    @memcpy(out_buf[0..parsed.branch.len], parsed.branch);
    return .{
        .branch = out_buf[0..parsed.branch.len],
        .dirty = parsed.dirty,
        .ahead = parsed.ahead,
        .behind = parsed.behind,
    };
}

const testing = std.testing;

test "parseStatus reads branch and dirty count" {
    const out =
        "## main...origin/main\n" ++
        " M src/a.zig\n" ++
        "?? new.txt\n";
    const info = parseStatus(out).?;
    try testing.expectEqualStrings("main", info.branch);
    try testing.expectEqual(@as(u32, 2), info.dirty);
    try testing.expectEqual(@as(u32, 0), info.ahead);
}

test "parseStatus reads ahead and behind" {
    const out = "## main...origin/main [ahead 3, behind 1]\n";
    const info = parseStatus(out).?;
    try testing.expectEqualStrings("main", info.branch);
    try testing.expectEqual(@as(u32, 3), info.ahead);
    try testing.expectEqual(@as(u32, 1), info.behind);
}

test "parseStatus handles a branch with no upstream" {
    const info = parseStatus("## feature/x\n").?;
    try testing.expectEqualStrings("feature/x", info.branch);
    try testing.expectEqual(@as(u32, 0), info.dirty);
}

test "parseStatus returns null without a branch header" {
    try testing.expect(parseStatus("") == null);
    try testing.expect(parseStatus("?? stray.txt\n") == null);
}

File: crates/anvil-prompt-core/src/git.rs
//! Git status for the prompt. `parse_status` (pure, tested) interprets the
//! output of `git status --porcelain=v1 --branch`; `query` runs git as a
//! subprocess and feeds it to the parser.

use std::path::Path;
use std::process::Command;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Info {
    pub branch: String,
    pub dirty: u32,
    pub ahead: u32,
    pub behind: u32,
}

/// Parse `git status --porcelain=v1 --branch` output.
/// Returns `None` if no branch header line is present.
pub fn parse_status(text: &str) -> Option<Info> {
    let mut info: Option<Info> = None;
    for line in text.split('\n') {
        if line.is_empty() {
            continue;
        }
        if let Some(rest) = line.strip_prefix("## ") {
            info = Some(parse_branch_line(rest));
        } else if let Some(i) = info.as_mut() {
            i.dirty += 1;
        }
    }
    info
}

fn parse_branch_line(rest: &str) -> Info {
    // e.g. "main...origin/main [ahead 1, behind 2]"  or  "main"
    let branch_end = {
        let by_dots = rest.find("...").unwrap_or(rest.len());
        let by_space = rest.find(' ').unwrap_or(rest.len());
        by_dots.min(by_space)
    };
    let branch = rest[..branch_end].to_string();

    let ahead = rest
        .find("ahead ")
        .map(|i| read_num(&rest[i + 6..]))
        .unwrap_or(0);
    let behind = rest
        .find("behind ")
        .map(|i| read_num(&rest[i + 7..]))
        .unwrap_or(0);

    Info {
        branch,
        dirty: 0,
        ahead,
        behind,
    }
}

fn read_num(s: &str) -> u32 {
    let digits: String = s.chars().take_while(|c| c.is_ascii_digit()).collect();
    digits.parse().unwrap_or(0)
}

/// Run git in `cwd` and return its status, or `None` if not a repo / git
/// fails / it errors.
pub fn query(cwd: &Path) -> Option<Info> {
    let output = Command::new("git")
        .args(["status", "--porcelain=v1", "--branch"])
        .current_dir(cwd)
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    parse_status(&stdout)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_status_reads_branch_and_dirty_count() {
        let out = "## main...origin/main\n M src/a.zig\n?? new.txt\n";
        let info = parse_status(out).unwrap();
        assert_eq!(info.branch, "main");
        assert_eq!(info.dirty, 2);
        assert_eq!(info.ahead, 0);
    }

    #[test]
    fn parse_status_reads_ahead_and_behind() {
        let out = "## main...origin/main [ahead 3, behind 1]\n";
        let info = parse_status(out).unwrap();
        assert_eq!(info.branch, "main");
        assert_eq!(info.ahead, 3);
        assert_eq!(info.behind, 1);
    }

    #[test]
    fn parse_status_handles_a_branch_with_no_upstream() {
        let info = parse_status("## feature/x\n").unwrap();
        assert_eq!(info.branch, "feature/x");
        assert_eq!(info.dirty, 0);
    }

    #[test]
    fn parse_status_returns_none_without_a_branch_header() {
        assert!(parse_status("").is_none());
        assert!(parse_status("?? stray.txt\n").is_none());
    }

    // ── Additional parse_branch_line coverage ──────────────────────────────────

    #[test]
    fn parse_branch_line_ahead_only() {
        let info = parse_status("## main...origin/main [ahead 2]\n").unwrap();
        assert_eq!(info.ahead, 2);
        assert_eq!(info.behind, 0);
    }

    #[test]
    fn parse_branch_line_behind_only() {
        let info = parse_status("## main...origin/main [behind 4]\n").unwrap();
        assert_eq!(info.ahead, 0);
        assert_eq!(info.behind, 4);
    }

    #[test]
    fn parse_branch_line_no_dots_no_upstream() {
        // Branch name with a space would end at the space boundary.
        let info = parse_status("## mybranch\n").unwrap();
        assert_eq!(info.branch, "mybranch");
        assert_eq!(info.ahead, 0);
        assert_eq!(info.behind, 0);
    }

    #[test]
    fn parse_status_multiple_dirty_lines() {
        let out = "## main\n M file1\n?? file2\n?? file3\n";
        let info = parse_status(out).unwrap();
        assert_eq!(info.dirty, 3);
    }

    #[test]
    fn read_num_non_digit_prefix_returns_zero() {
        // read_num is private; test via parse_status with malformed ahead value.
        // "ahead abc" → read_num("abc") → 0
        let info = parse_status("## main...origin/main [ahead abc]\n").unwrap();
        assert_eq!(info.ahead, 0);
    }

    // ── query: runs git in a temp dir (integration test) ──────────────────────

    #[test]
    fn query_non_repo_returns_none() {
        // /tmp should not be a git repo.
        let result = query(std::path::Path::new("/tmp"));
        // In most cases this returns None (not a git repo).
        // If /tmp is inside a git repo on the CI runner, skip the assertion.
        let _ = result; // Either None or Some is acceptable here.
    }

    #[test]
    fn query_current_dir_returns_some_or_none() {
        // The test runner's cwd might be in the anvil repo.
        if let Ok(cwd) = std::env::current_dir() {
            let result = query(&cwd);
            // We don't assert the value — just that it doesn't panic.
            let _ = result;
        }
    }
}
